---
title: "01g_mitofish-filter"
date: "2024-02-26"
---

This script generates a filter for the Mitofish fasta.
Input: .csv with accession and description from NCBI
output: .txt with List of accessions to keep.

```{r setup, include=FALSE}
library(tidyverse)
library(here)

# here filepaths
crabsdb <- here('database12s', 'crabs')

datefield <- '20241010_'
```


```{r loadin}
mitofishdb <- read_csv(here(crabsdb, '20240226_mitofish_ncbi-taxonomy.csv')) %>%
  select(!desc_verbatim)

# if updating
mitofishdb2 <- read_csv(here(crabsdb, '20241010_mitofish_ncbi-taxonomy.csv')) %>%
  select(!desc_verbatim)

mitofishdb <- mitofishdb %>%
  bind_rows(mitofishdb2) %>%
  distinct()
```

```{r}
ncbi_fish_gene <- mitofishdb %>%
  # remove decimals on accession
  mutate(accession = str_remove_all(accession, "\\.[1-9]$")) %>%
  # just in case
  distinct() %>%
  mutate(unverified = ifelse(str_detect(description, 'UNVERIFIED'), "yes", "no")) %>%
  mutate(genome = ifelse(str_detect(description, 'complete genome|partial genome|partiale genome|mitochondrion sequence|mitochondrial DNA|genome assembly|mitochondrial genome|mitochondrial sequence|mitochondrionial geneome|mitochondrial genes'), 
                         "yes", "no")) %>%
  mutate(small_subunit = ifelse(str_detect(description, '(small subunit)'), 'yes', 'no')) %>%
  mutate(has_12s = ifelse(str_detect(description, '( 12s)|( 12S)|12 ribosomal|12S ribosomal'), "yes", "no")) %>%
  mutate(has_16s = ifelse(str_detect(description, '16s|16S|16 ribosomal'), "yes", "no")) %>%
  mutate(has_18s = ifelse(str_detect(description, '18s|18S|18 ribosomal'), "yes", "no")) %>%
  mutate(has_28s = ifelse(str_detect(description, '28s|28S|28 ribosomal'), "yes", "no")) %>%
  mutate(has_cytb = ifelse(str_detect(description, 'cytb|cybt|(cyt b)|(Cyt b)|Cytb|CytB|CYTB|cytochrome b|Cytochrome b|cytochrome oxidase b|cytochrome B|cytochorome B|cytochorome b|cytothrome b|cytochrome-b'), 
                           "yes", "no")) %>%
  mutate(has_cytc = ifelse(str_detect(description, 'cytc|(cyt c)|CYTC|cytochrome c|cytochrome oxidase subunit 3|cytochrome oxidase c'), "yes", "no")) %>%
  mutate(has_cob = ifelse(str_detect(description, 'apocytochrome'), "yes", "no")) %>%
  mutate(has_coi = ifelse(str_detect(description, 
                                     'coi|co1|COI|CO1|cox1|COX1|COXI|coxi|coxI|cox1|cytochrome oxidase subunit I|cytochrome oxidase subunit 1|cytochrome oxidase I|cytochome oxidase subunit I'), 
                          "yes", "no")) %>%
  mutate(has_rrna = ifelse(str_detect(description, 'subunit ribosomal RNA|small subunit rRNA|large subunit rRNA|ribosomal RNA large subunit|ribosomal protein|large ribosomal RNA'), 
                           "yes", "no")) %>%
  mutate(has_nadh = ifelse(str_detect(description, 'NADH dehydrogenase|NADH dehdyrogenase|NADH-ubiquinone oxidoreductase|NADH1|NADH subunit'), 
                           "yes", "no")) %>%
  mutate(has_atp = ifelse(str_detect(description, 'ATPase|ATP synthase|ATP synthetase|ATP gene|atp6|atp8|ATP6|ATP8'), "yes", "no")) %>%
  mutate(has_ade = ifelse(str_detect(description, 'adenosinetriphosphatase'), "yes", "no")) %>%
  mutate(has_nico = ifelse(str_detect(description, 'nicotinamide adenine dehydrogenase'), "yes", "no")) %>%
  mutate(has_ctrl = ifelse(str_detect(description, 'control region'), "yes", "no")) %>%
  mutate(has_nd = ifelse(str_detect(description, 'ND4|ND2|ND5|ND6'), "yes", "no")) %>%
  mutate(has_close_trna = ifelse(str_detect(description, 'tRNA-Val|tRNA-Phe'), 
                               "yes", "no")) %>%
  mutate(has_trna = ifelse(str_detect(description, 'tRNA-His|tRNA-Ser|tRNA-Leu|tRNA-Val|tRNA-Gln|tRNA-Lys|tRNA-Pro|tRNA-Phe|tRNA-Thr|tRNA-Glu|tRNA-Asn'), 
                               "yes", "no")) %>%
  # note dloop has some with "except for dloop"
  mutate(has_dloop = ifelse(str_detect(description, 'D-loop|dloop'), "yes", "no")) %>%
  mutate(has_dloop = ifelse(str_detect(description, 'except for D-loop'), "no", has_dloop)) %>%
  mutate(has_other = ifelse(str_detect(description, 'rps7|HP54|s7|S7|dehydrogenase subunit two|'), "yes", "no")) %>%
  mutate(has_noncode = ifelse(str_detect(description, 'non-coding'), "yes", "no")) %>%
  mutate(all_nos = ifelse(genome == "no" & 
                            has_12s == "no" &
                            has_16s == "no" &
                            has_18s == "no" &
                            has_28s == "no" &
                            has_cytb == "no" &
                            has_cytc == "no" &
                            has_cob == "no" &
                            has_coi == "no" &
                            has_rrna == "no" &
                            has_nadh == "no" &
                            has_atp == "no" &
                            has_ade == "no" &
                            has_nico == "no" &
                            has_ctrl == "no" &
                            has_nd == "no" &
                            has_trna == "no" &
                            has_noncode == "no" &
                            has_other == "no" &
                            has_dloop == "no", "yes", "no"))

write_csv(ncbi_fish_gene, here(crabsdb, paste0(datefield, 'mitofish_by-gene.csv')))

# separate the dirty column to clean
ncbi_organism <- ncbi_fish_gene %>%
  filter(genome == 'yes' | small_subunit == 'yes' | has_12s == 'yes' | has_rrna == 'yes') %>%
  #select(description) %>%
  distinct()
```

```{r write list}
#make list
ncbi_accessions <- ncbi_organism %>%
  select(accession) %>%
  distinct()

write_delim(ncbi_accessions, here(crabsdb, paste0(datefield, 'mitofish_12s.txt')),
            col_names = FALSE)

```

